### ISSUE TITLE
Getting sample index from the reader

### ISSUE THREAD

**Issue by vturrisi** [NVIDIA/DALI#3238]
Hey,
We need the sample indexes for some computations, but we didn't manage to get them with any workaround that we tried.

Based on this [1759](https://github.com/NVIDIA/DALI/issues/1759), we tried doing bit shifting to store both the class label and the sample index, decoding both on a custom `DALIGenericIterator`. Also, we tried using the sample indexes as "labels" for the `reader` and then getting the actual class labels using a lookup table that maps sample indexes to class labels.
For both we are having the same issues, so I'll just focus on the second one.

This is how my iterator looks like:
```python
class PretrainWrapper(DALIGenericIterator):
    def __init__(
        self,
        model_batch_size: int,
        model_rank: int,
        model_device: str,
        conversion_map: list,
        *args,
        **kwargs,
    ):
        super().__init__(*args, **kwargs)
        self.model_batch_size = model_batch_size
        self.model_rank = model_rank
        self.model_device = model_device
        self.conversion_map = conversion_map
        if self.conversion_map is not None:
            self.conversion_map = torch.tensor(
                self.conversion_map, dtype=torch.float32, device=self.model_device
            ).reshape(-1, 1)
            self.conversion_map = nn.Embedding.from_pretrained(self.conversion_map)

    def __next__(self):
        batch = super().__next__()[0]
        if self.conversion_map is not None:
            *all_X, indexes = [batch[v] for v in self.output_map]
            targets = self.conversion_map(indexes).detach().flatten().long()
            indexes = indexes.flatten().long()
            print('inside wrapper', "targets:", targets[:3], "indexes:", indexes[:3])
        else:
            *all_X, targets = [batch[v] for v in self.output_map]
            targets = targets.squeeze(-1).long()
            # creates dummy indexes
            indexes = torch.arange(self.model_batch_size, device=self.model_device) + (
                self.model_rank * self.model_batch_size
            )

        return indexes, all_X, targets
```
I'm using `pytorch lightning`, so this is what I'm doing in the `training_step` of my model, for debugging:
```python
def training_step(self, batch: List[Any], batch_idx: int) -> Dict[str, Any]:
    indexes, X, targets = batch

    print("on training step", "targets:", targets[:3], "indexes:", indexes[:3])
    exit()
```

When I run, my output looks like this:
```
Epoch 0:   0%|                                                                                                                                           | 0/1029 [00:00<?, ?it/s]
inside wrapper targets: tensor([66, 73, 59], device='cuda:0') indexes: tensor([84272, 92729, 75867], device='cuda:0')
inside wrapper targets: tensor([74, 21, 69], device='cuda:0') indexes: tensor([94199, 26693, 88216], device='cuda:0')
on training step targets: tensor([66, 73, 59], device='cuda:0') indexes: tensor([94199, 26693, 88216], device='cuda:0')
```
I assume that there's a prefetch step and that's why I see two prints "inside wrapper", but I can't figure out why my data is getting mixed (targets come from the last call while indexes come from the current call). Note that targets are generated inside the Wrapper, while indexes come directly from a `DALI` pipeline.

Tried my best to describe what we are experiencing, but I can provide more info if needed :)

To reproduce:

- This branch has the code in the right setup: https://github.com/vturrisi/solo-learn/tree/dali_issue

* git clone -b dali_issue git@github.com:vturrisi/solo-learn.git
* cd bash_files/pretrain/imagenet-100
* modify barlow.sh to point to any dataset in the normal pytorch format via the following flags:
```
--data_dir /datasets \
--train_dir imagenet-100/train \
--val_dir imagenet-100/val \
```
* bash barlow.sh

**Comment by JanuszL** [comment]
Hi @vturrisi,

I believe this may be the cause https://github.com/PyTorchLightning/pytorch-lightning/issues/1316 - PyTorch lighting does own buffering while DALI does own the PyTorch tensors it returns. In result the content of tensors is trashed.
Please check how DALI is integrated with NEMO - https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/data/audio_to_text_dali.py#L450L454 and how it deals with this problem.

**Comment by vturrisi** [comment]
Hey @JanuszL 

I see. I also saw that you modified the usage example of dali for lightning. Doing .detach().clone() seems to have fixed my problem. Thank you!

**Comment by JanuszL** [comment]
Yes, I did it to avoid such problems in the future.
